### PR TITLE
Put model in eval mode for evaluation

### LINKED ISSUE #1047: Model eval
Does the elether_eval recipe call model.eval() somewhere? I tried to find it but could not. It's there in the lm evaluation harness but we have overriden the constructor in eval wrapper. Also, my results change when I switch model to eval mode.

### PR DESCRIPTION
**Pull request by joecummings**
#### Context

This is just best practices. Model should be put in eval mode during inference for evaluation.

**How did we not notice this before?** [model.eval](https://pytorch.org/docs/stable/generated/torch.nn.Module.html#torch.nn.Module.eval) only acts on specific layers that act different between training and eval, like `Dropout`. We didn't test any modules with dropout so we missed it.

Closes #1047 

Caveat: SPDA dropout is not able to be turned off by `model.eval()` as per https://github.com/pytorch/pytorch/issues/124464

#### Changelog
* Adds `model.eval()`

#### Test plan

Expect same results with our models since we don't apply dropout anywhere.

<details>
<summary>With `model.eval()`</summary>

```bash
(joe-torchtune) [jrcummings@devgpu012.cln5 ~/projects/joe-torchtune (add-eval-decorator)]$ tune run eleuther_eval --config eleuther_evaluation
2024-07-29:12:25:22,515 INFO     [_utils.py:33] Running EleutherEvalRecipe with resolved config:

batch_size: 8
checkpointer:
  _component_: torchtune.utils.FullModelHFCheckpointer
  checkpoint_dir: /tmp/gemma
  checkpoint_files:
  - model-00001-of-00002.safetensors
  - model-00002-of-00002.safetensors
  model_type: GEMMA
  output_dir: /tmp/gemma
device: cuda
dtype: bf16
limit: null
max_seq_length: 4096
model:
  _component_: torchtune.models.gemma.gemma_2b
quantizer: null
seed: 1234
tasks:
- truthfulqa_mc2
tokenizer:
  _component_: torchtune.models.llama2.llama2_tokenizer
  path: /tmp/gemma/tokenizer.model

2024-07-29:12:25:30,259 DEBUG    [seed.py:60] Setting manual seed to local seed 1234. Local seed is seed + rank = 1234 + 0
2024-07-29:12:25:31,825 INFO     [eleuther_eval.py:228] Model is initialized with precision torch.bfloat16.
2024-07-29:12:25:31,959 INFO     [eleuther_eval.py:208] Tokenizer is initialized from file.
2024-07-29:12:25:33,465 INFO     [huggingface.py:162] Using device 'cuda:0'
/home/jrcummings/.conda/envs/joe-torchtune/lib/python3.11/site-packages/huggingface_hub/file_download.py:1132: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.
  warnings.warn(
2024-07-29:12:25:43,520 INFO     [eleuther_eval.py:253] Running evaluation on ['truthfulqa_mc2'] tasks.
2024-07-29:12:25:43,521 INFO     [task.py:395] Building contexts for truthfulqa_mc2 on rank 0...
100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████| 817/817 [00:01<00:00, 784.39it/s]
2024-07-29:12:25:44,624 INFO     [evaluator.py:362] Running loglikelihood requests
Running loglikelihood requests: 100%|████████████████████████████████████████████████████████████████████████████| 5882/5882 [00:46<00:00, 127.67it/s]
2024-07-29:12:26:35,502 INFO     [eleuther_eval.py:260] Eval completed in 63.38 seconds.
|    Tasks     |Version|Filter|n-shot|Metric|Value |   |Stderr|
|--------------|------:|------|-----:|------|-----:|---|-----:|
|truthfulqa_mc2|      2|none  |     0|acc   |0.3993|±  |0.0152|
```
</details>

<details>
<summary>Without `model.eval()`</summary>

```bash
joe-torchtune) [jrcummings@devgpu012.cln5 ~/projects/joe-torchtune (add-eval-decorator)]$ tune run eleuther_eval --config eleuther_evaluation
2024-07-29:12:28:37,446 INFO     [_utils.py:33] Running EleutherEvalRecipe with resolved config:

batch_size: 8
checkpointer:
  _component_: torchtune.utils.FullModelHFCheckpointer
  checkpoint_dir: /tmp/gemma
  checkpoint_files:
  - model-00001-of-00002.safetensors
  - model-00002-of-00002.safetensors
  model_type: GEMMA
  output_dir: /tmp/gemma
device: cuda
dtype: bf16
limit: null
max_seq_length: 4096
model:
  _component_: torchtune.models.gemma.gemma_2b
quantizer: null
seed: 1234
tasks:
- truthfulqa_mc2
tokenizer:
  _component_: torchtune.models.llama2.llama2_tokenizer
  path: /tmp/gemma/tokenizer.model

2024-07-29:12:28:41,374 DEBUG    [seed.py:60] Setting manual seed to local seed 1234. Local seed is seed + rank = 1234 + 0
2024-07-29:12:28:45,231 INFO     [eleuther_eval.py:228] Model is initialized with precision torch.bfloat16.
2024-07-29:12:28:45,356 INFO     [eleuther_eval.py:208] Tokenizer is initialized from file.
2024-07-29:12:28:46,872 INFO     [huggingface.py:162] Using device 'cuda:0'
/home/jrcummings/.conda/envs/joe-torchtune/lib/python3.11/site-packages/huggingface_hub/file_download.py:1132: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.
  warnings.warn(
2024-07-29:12:28:56,605 INFO     [eleuther_eval.py:253] Running evaluation on ['truthfulqa_mc2'] tasks.
2024-07-29:12:28:56,606 INFO     [task.py:395] Building contexts for truthfulqa_mc2 on rank 0...
100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████| 817/817 [00:01<00:00, 765.82it/s]
2024-07-29:12:28:57,734 INFO     [evaluator.py:362] Running loglikelihood requests
Running loglikelihood requests: 100%|████████████████████████████████████████████████████████████████████████████| 5882/5882 [00:45<00:00, 130.47it/s]
2024-07-29:12:29:47,759 INFO     [eleuther_eval.py:260] Eval completed in 62.26 seconds.
|    Tasks     |Version|Filter|n-shot|Metric|Value |   |Stderr|
|--------------|------:|------|-----:|------|-----:|---|-----:|
|truthfulqa_mc2|      2|none  |     0|acc   |0.3993|±  |0.0152|

```

</details>

### CODE REVIEW

**Review by felipemello1 — APPROVED** [review]

**Comment by felipemello1** [comment]
@joecummings , should we change everywhere we have scaled_dot_product to do this check: https://github.com/pytorch/pytorch/issues/124464#issuecomment-2073355439

`F.scaled_dot_product_attention(..., dropout_p=(self.p if self.training else 0.0))`

**Comment by joecummings** [comment]
> @joecummings , should we change everywhere we have scaled_dot_product to do this check: [pytorch/pytorch#124464 (comment)](https://github.com/pytorch/pytorch/issues/124464#issuecomment-2073355439)
> 
> `F.scaled_dot_product_attention(..., dropout_p=(self.p if self.training else 0.0))`

This is a more involved change since we don't even define a `self.training` attr right now, but yes that might be a good idea. We are currently getting away with it b/c the only dropout used here is during training for LoRA, which we then merge back into the main model which sets dropout back to zero.

**Comment by felipemello1** [comment]
@joecummings that doesnt sound like a priority, but maybe an issue for the backlog?
